🐛clustercatalog_controller: hacky, but more correct status updating in reconciler

joelanford opened the pull request:
For awhile in catalogd, we've had a very loose "needsUnpack" behavior where we would make the assumption that if we didn't need to unpack, then we _also_ didn't need to update the status of the object.

This means that some objects may _never_ have their status updated after an upgrade. That's problematic because upgrades may add new helpful fields or fix bugs in the way that status is reported.

This PR is a somewhat hacky fix for the situation. It solves the problem by keeping an in-memory map of the observedGeneration and resolved source, and then using _that_ as the source of truth to and then use that expected status as part of the determination of whether an unpack is required.

A new criteria of "needsUnpack" is now: "does the expected status differ from the status that is on the object. If so, unpack again.

Fixes #422 
Fixes #420 

<!--
Please prefix the title of this PR with one of the following icons:

    * ⚠ (:warning:, major/breaking change)
    * ✨ (:sparkles:, minor/compatible change)
    * 🐛 (:bug:, patch/bug fix)
    * 📖 (:book:, docs)
    * 🌱 (:seedling:, other)

-->

joelanford commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -47,9 +50,22 @@ const (
+	Unpacker source.Unpacker
+	Storage  storage.Instance
+
+	finalizers crfinalizer.Finalizers
```
I needed to refactor finalizers so that it had access to the reconciler struct. That way it would be possible to `delete(storedCatalogs, catalog.Name)` in the finalizer logic.

joelanford commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -133,55 +159,125 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
+	case r.needsPoll(storedCatalog.unpackResult.ResolvedSource.Image.LastSuccessfulPollAttempt.Time, catalog):
+		l.Info("unpack required: poll duration has elapsed")
+		needsUnpack = true
+	}
```
I intentionally added lots of logging here to help us deduce why we might be unpacking more frequently than expected.

joelanford commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -133,55 +159,125 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
+	clearUnknownConditions(expectedStatus)
+	if hasStoredCatalog && r.Storage.ContentExists(catalog.Name) {
+		updateStatusServing(expectedStatus, storedCatalog.unpackResult, r.Storage.ContentURL(catalog.Name), storedCatalog.observedGeneration)
+		updateStatusProgressing(expectedStatus, storedCatalog.observedGeneration, nil)
```
I needed to change the `updateStatusProgressing` signature so that I could call it with a deep-copied status, and a different observed generation.

bentito commented on internal/source/containers_image_test.go:
```diff
@@ -396,7 +396,7 @@ func TestImageRegistry(t *testing.T) {
 				if tt.digestAlreadyExists {
 					assert.Contains(t, buf.String(), "image already unpacked")
-					assert.Equal(t, rs.UnpackTime, unpackDirStat.ModTime())
+					assert.Equal(t, rs.UnpackTime, unpackDirStat.ModTime().Truncate(time.Second))
```
I think even asserting on the second could be brittle if we happen to run on the second to next second break. How about just asserting the duration is in a reasonable range?
```suggestion
					assert.WithinDuration(t, rs.UnpackTime, unpackDirStat.ModTime(), time.Second)
```

joelanford replied:
This one should be exact. `rs.UnpackTime` is always truncated to the second in the same way.  I'll add a comment in the code about why we need that though.

bentito commented:
@joelanford when I point current operator-controller main at this with a `replace` and run `make test-upgrade-e2e` I am still seeing:

```bash
# github.com/operator-framework/operator-controller/internal/catalogmetadata/cache
internal/catalogmetadata/cache/cache.go:86:42: catalog.Status.ResolvedSource.Image.ResolvedRef undefined (type *"github.com/operator-framework/catalogd/api/core/v1alpha1".ResolvedImageSource has no field or method ResolvedRef)
internal/catalogmetadata/cache/cache.go:122:62: catalog.Status.ResolvedSource.Image.ResolvedRef undefined (type *"github.com/operator-framework/catalogd/api/core/v1alpha1".ResolvedImageSource has no field or method ResolvedRef)
internal/catalogmetadata/cache/cache.go:164:52: catalog.Status.ResolvedSource.Image.ResolvedRef undefined (type *"github.com/operator-framework/catalogd/api/core/v1alpha1".ResolvedImageSource has no field or method ResolvedRef)
# github.com/operator-framework/operator-controller/internal/controllers
internal/controllers/clusterextension_controller.go:419:53: oldObject.Status.ResolvedSource.Image.ResolvedRef undefined (type *"github.com/operator-framework/catalogd/api/core/v1alpha1".ResolvedImageSource has no field or method ResolvedRef)
internal/controllers/clusterextension_controller.go:419:106: newObject.Status.ResolvedSource.Image.ResolvedRef undefined (type *"github.com/operator-framework/catalogd/api/core/v1alpha1".ResolvedImageSource has no field or method ResolvedRef)
# github.com/operator-framework/operator-controller/internal/controllers [github.com/operator-framework/operator-controller/internal/controllers.test]
internal/controllers/clusterextension_controller.go:419:53: oldObject.Status.ResolvedSource.Image.ResolvedRef undefined (type *"github.com/operator-framework/catalogd/api/core/v1alpha1".ResolvedImageSource has no field or method ResolvedRef)
internal/controllers/clusterextension_controller.go:419:106: newObject.Status.ResolvedSource.Image.ResolvedRef undefined (type *"github.com/operator-framework/catalogd/api/core/v1alpha1".ResolvedImageSource has no field or method ResolvedRef)
make: *** [vet] Error 1
```

joelanford commented:
@bentito, that's expected. See the accompanying changes in https://github.com/operator-framework/operator-controller/pull/1329.

No matter how you slice it, there is a breaking change from 0.29.0 to 0.30.0 (and from 0.29.0 to this PR):
1. We removed some status fields that operator-controller depends on, and
2. We changed the `spec.source.type` enum from `image` to `Image`, with no backcompat support. So the upgrade tests are bound to fail because the new catalogd no longer understands old data (i.e. `spec.source.type: "image"`)

bentito commented:
> @bentito, that's expected. See the accompanying changes in [operator-framework/operator-controller#1329](https://github.com/operator-framework/operator-controller/pull/1329).
> 
> No matter how you slice it, there is a breaking change from 0.29.0 to 0.30.0 (and from 0.29.0 to this PR):
> 
>     1. We removed some status fields that operator-controller depends on, and
> 
>     2. We changed the `spec.source.type` enum from `image` to `Image`, with no backcompat support. So the upgrade tests are bound to fail because the new catalogd no longer understands old data (i.e. `spec.source.type: "image"`)

ah, okay, understanding that now, was just worried it was supposed to be fixed from catalogd side alone

ankitathomas reviewed (approved):
Looks good! Left a small comment about the status comparison

joelanford commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -133,55 +159,125 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
+	clearUnknownConditions(expectedStatus)
+	if hasStoredCatalog && r.Storage.ContentExists(catalog.Name) {
+		updateStatusServing(expectedStatus, storedCatalog.unpackResult, r.Storage.ContentURL(catalog.Name), storedCatalog.observedGeneration)
+		updateStatusProgressing(expectedStatus, storedCatalog.observedGeneration, nil)
```
Oh this is a good question. Here's a scenario:
1. We are successfully serving content for generation 1
2. We reconcile generation 2, which has a bad image ref, let's say the tag doesn't exist. So we:
    - Do not update `storedCatalog` map
    - DO update the status with `Progressing=True, Retrying`
3. On the next reconcile (that is requeued basically immediately), we build an "expectedStatus" that says `Installed=True` and `Progressing=False` based on this code. We later compare this to the real status from the object and we are told that there are differences, which forces an unpack.

Step 3 is a little odd though. Maybe if we return an error from `Reconcile` we should `delete(r.storedCatalogs, catalog.Name)`?
